File: exercism/zig/armstrong-numbers/armstrong_numbers.zig
pub fn isArmstrongNumber(num: u128) bool {
    var digits: u32 = 1;
    var x = num;
    while (x > 9) : (x /= 10) digits += 1;
    var sum: u128 = 0;
    x = num;
    while (x != 0) : (x /= 10)
        sum += @import("std").math.powi(u128, x % 10, digits) catch unreachable;
    return sum == num;
}

File: exercism/c/armstrong-numbers/armstrong_numbers.c
#include "armstrong_numbers.h"

bool is_armstrong_number(int candidate)
{
    int digits = 1;
    int x = candidate;
    while (x > 9 || x < -9) {
        x /= 10;
        ++digits;
    }

    int r = 0;
    x = candidate;
    while (x) {
        int d = 1;
        for (int i = 0; i < digits; ++i)
            d *= x % 10;
        r += d;
        x /= 10;
    }

    return r == candidate;
}
